integration/helpers/instance.go

```diff
@@ -1313,6 +1334,205 @@ func (i *TeleInstance) NewClient(cfg ClientConfig) (*client.TeleportClient, erro
+		User: user,
+		Pass: password,
+	})
+	if err != nil {
+		return nil, nil, trace.Wrap(err)
+	}
+
+	// Create POST request to create session.
+	u := url.URL{
+		Scheme: "https",
+		Host:   proxyHost,
+		Path:   "/v1/webapi/sessions/web",
+	}
+	req, err := http.NewRequest(http.MethodPost, u.String(), bytes.NewBuffer(csReq))
+	if err != nil {
+		return nil, nil, trace.Wrap(err)
+	}
+
+	// Attach CSRF token in cookie and header.
+	csrfToken, err := utils.CryptoRandomHex(32)
+	if err != nil {
+		return nil, nil, trace.Wrap(err)
+	}
+
+	req.AddCookie(&http.Cookie{
+		Name:  csrf.CookieName,
+		Value: csrfToken,
+	})
+	req.Header.Set("Content-Type", "application/json; charset=utf-8")
+	req.Header.Set(csrf.HeaderName, csrfToken)
+
+	// Issue request.
+	httpClient := &http.Client{
+		Transport: &http.Transport{
+			TLSClientConfig: &tls.Config{
+				InsecureSkipVerify: true,
+			},
+		},
+	}
+	resp, err := httpClient.Do(req)
+	if err != nil {
+		return nil, nil, trace.Wrap(err)
+	}
+
+	defer resp.Body.Close()
+	if http.StatusOK != resp.StatusCode {
+		return nil, nil, trace.ConnectionProblem(nil, "received unexpected status code: %d", resp.StatusCode)
+	}
+
+	// Read in response.
+	var csResp *web.CreateSessionResponse
+	err = json.NewDecoder(resp.Body).Decode(&csResp)
+	if err != nil {
+		return nil, nil, trace.Wrap(err)
+	}
+
+	return csResp, resp.Cookies(), nil
+}
+
+// SSH establishes an SSH connection via the web api in the same manner that
+// the web UI does. The returned [web.TerminalStream] should be used as stdin/stdout
+// for the session.
+func (w *WebClient) SSH(termReq web.TerminalRequest) (*web.TerminalStream, error) {
+	u := url.URL{
+		Host:   w.i.Web,
+		Scheme: client.WSS,
+		Path:   fmt.Sprintf("/v1/webapi/sites/%v/connect", w.tc.SiteName),
+	}
+	data, err := json.Marshal(termReq)
+	if err != nil {
+		return nil, err
+	}
+
+	q := u.Query()
+	q.Set("params", string(data))
+	q.Set(roundtrip.AccessTokenQueryParam, w.token)
+	u.RawQuery = q.Encode()
+
+	header := http.Header{}
+	header.Add("Origin", "http://localhost")
+	for _, cookie := range w.cookies {
+		header.Add("Cookie", cookie.String())
+	}
+
+	dialer := websocket.Dialer{
+		TLSClientConfig: &tls.Config{
+			InsecureSkipVerify: true,
+		},
+	}
+	ws, resp, err := dialer.Dial(u.String(), header)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	ty, raw, err := ws.ReadMessage()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if ty != websocket.BinaryMessage {
+		return nil, trace.BadParameter("unexpected websocket message; got %d want %d", ty, websocket.BinaryMessage)
+	}
+
+	var env web.Envelope
+	err = proto.Unmarshal(raw, &env)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	type siteSessionGenerateResponse struct {
+		Session session.Session `json:"session"`
+	}
+
+	var sessResp siteSessionGenerateResponse
+	err = json.Unmarshal([]byte(env.Payload), &sessResp)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	err = resp.Body.Close()
```

Comment:
Consider calling this in a defer so that it doesn't leak if any of the errors between dialer.Dial() and this line are hit.